Python refactor 2

zrax opened the pull request:
This refactor shouldn't have too many behavior changes, but significantly reduces the boilerplate and complexity for adding new bindings.
* Use macros for declaring trivial property get/set methods
* Use macros for declaring (and in some cases defining) common methods, including the signature cast.
  * This means the function declarations are now type-matched and don't need to be manually cast in the type tables.
* Use macros for method declarations, which forces consistency with the method flags and method declaration, and also places the documentation strings adjacent to the method itself.
* Use a macro for adding constants to a class declaration
* **DROP TABLES** -- that is, replace the large tables full of mostly nulls with (mostly) empty macro-initialized tables, and initialize the fields in the init function instead.
  * Side note: This means it's no longer unreasonable to place multiple class bindings for trivial classes in the same source file.

This set does NOT move all the get/set properties to using the new macros yet.  I used those as sentinels for properties that I needed to revisit, which will come in python_refactor_3 (the part with potentially breaking changes).  Most notably, most properties which deal with lists/tuples are left out of this PR.

Hoikas commented:
Compilation errors on MSVC++2015:
```
Error	C2910	'pyPlasma_get' : cannot be explicitly specialized	PyHSPlasma	d:\plasma\libhsplasma\python\PRP\Audio\pySoundBuffer.h	35	
Error	C2989	'plWAVHeader' : class template has already been declared as a non-class template	PyHSPlasma	d:\plasma\libhsplasma\python\prp\audio\pySoundBuffer.h	36	
Error	C2910	'pyPlasma_get' : cannot be explicitly specialized	PyHSPlasma	d:\plasma\libhsplasma\python\prp\audio\pySoundBuffer.h	35	
Error	C1903	unable to recover from previous error(s); stopping compilation	PyHSPlasma	d:\plasma\libhsplasma\python\prp\audio\pySoundBuffer.h	38	
Error	C2989	'plWAVHeader' : class template has already been declared as a non-class template	PyHSPlasma	d:\plasma\libhsplasma\python\prp\audio\pySoundBuffer.h	36	
Error	C2910	'pyPlasma_get' : cannot be explicitly specialized	PyHSPlasma	d:\plasma\libhsplasma\python\prp\audio\pySoundBuffer.h	35	
Error	C1903	unable to recover from previous error(s); stopping compilation	PyHSPlasma	d:\plasma\libhsplasma\python\prp\audio\pySoundBuffer.h	38	
Error	C2440	'initializing' : cannot convert from 'hashfunc' to 'long (__cdecl *)(PyObject *)'	PyHSPlasma	D:\Plasma\libhsplasma\Python\PRP\KeyedObject\pyKey.cpp	34	
Error	C2440	'=' : cannot convert from 'long (__cdecl *)(PyObject *)' to 'hashfunc'	PyHSPlasma	D:\Plasma\libhsplasma\Python\PRP\KeyedObject\pyKey.cpp	276	
Error	C2440	'initializing' : cannot convert from 'hashfunc' to 'long (__cdecl *)(PyObject *)'	PyHSPlasma	D:\Plasma\libhsplasma\Python\PRP\KeyedObject\pyLocation.cpp	44	
Error	C2440	'=' : cannot convert from 'long (__cdecl *)(PyObject *)' to 'hashfunc'	PyHSPlasma	D:\Plasma\libhsplasma\Python\PRP\KeyedObject\pyLocation.cpp	204	
Error	C2989	'plWAVHeader' : class template has already been declared as a non-class template	PyHSPlasma	D:\Plasma\libhsplasma\Python\PRP\Audio\pySoundBuffer.h	36	
Error	C2910	'pyPlasma_get' : cannot be explicitly specialized	PyHSPlasma	D:\Plasma\libhsplasma\Python\PRP\Audio\pySoundBuffer.h	35	
Error	C1903	unable to recover from previous error(s); stopping compilation	PyHSPlasma	D:\Plasma\libhsplasma\Python\PRP\Audio\pySoundBuffer.h	38	
Error	C2989	'plWAVHeader' : class template has already been declared as a non-class template	PyHSPlasma	d:\plasma\libhsplasma\python\PRP\Audio\pySoundBuffer.h	36	
```

Deledrius commented:
@Hoikas, the plWaveHeader error is the one I reported in IRC last week.  There's a [redundant class declaration](https://github.com/H-uru/libhsplasma/pull/85/files#diff-2d20a7509aebd5b87e597005f56c00dbR36).

The two hashfunc <-> long conversion errors from pyKey and pyLocation are new though.

zrax commented:
It's not so much a redundant class declaration as a bug in MSVC's parser -- it's treating `template [...] class plWAVHeader* [...]` as a class declaration for plWAVHeader, rather than interpreting `class` as the type of the `plWAVHeader` return value :(.  Either way, it should be fixed now.

Deledrius commented:
Oh, wow.  I wasn't aware that was a valid return value.  Thanks for explaining. 😸

Hoikas commented:
`plCoordinateInterface` appears to no longer have the member function `ClassIndex`. Don't know if this applies to the other KOs

zrax commented:
Looks like `plCreatable` was initialized after some of its subtypes...  This worked before since most of the structure was declared statically, but doesn't any more.  I added a check and fixed both cases.

Hoikas commented on Python/PRP/Modifier/pyPythonParameter.cpp:
```diff
@@ -86,120 +86,98 @@ static PyObject* pyPythonParameter_write(pyPythonParameter* self, PyObject* args
-static PyObject* pyPythonParameter_getType(pyPythonParameter* self, void*) {
-    return PyInt_FromLong(self->fThis->fValueType);
-}
+PY_PROPERTY_MEMBER(unsigned int, PythonParameter, id, fID)
```
I'm getting an access violation on this setter in Korman

zrax replied:
Fixed

Deledrius commented on Python/PRP/Animation/pyViewFaceModifier.cpp:
```diff
@@ -26,196 +26,46 @@
-}
+PY_PROPERTY(hsVector3, ViewFaceModifier, scale, getScale, setScale)
+PY_PROPERTY(hsVector3, ViewFaceModifier, offset, getOffset, setOffset)
+PY_PROPERTY(hsMatrix44, ViewFaceModifier, localToparent, getLocalToParent, setLocalToParent)
```
Typo here.  Name should be localToParent.

Deledrius commented on Python/PRP/Animation/pyViewFaceModifier.cpp:
```diff
@@ -26,196 +26,46 @@
-    { NULL, NULL, NULL, NULL, NULL }
+    pyViewFaceModifier_scale_getset,
+    pyViewFaceModifier_offset_getset,
+    pyViewFaceModifier_localToparent_getset,
```
Typo, same as above.
